classes/observers/events/PKPUsageEvent.inc.php

```diff
@@ -0,0 +1,204 @@
+    public function __construct(int $assocType, int $assocId, int $contextId, int $submissionId = null, int $representationId = null, string $mimetype = null)
+    {
+        $application = Application::get();
+        $request = $application->getRequest();
+
+        $time = Core::getCurrentDate();
+
+        $ip = $request->getRemoteAddr();
+
+        $canonicalUrlPage = $canonicalUrlOp = null;
+        $canonicalUrlParams = [];
+
+        switch ($assocType) {
+            case Application::ASSOC_TYPE_SUBMISSION_FILE_COUNTER_OTHER:
+            case Application::ASSOC_TYPE_SUBMISSION_FILE:
+                $canonicalUrlOp = 'download';
+                $canonicalUrlParams = [$submissionId];
+                $router = $request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($request);
+                $args = $router->getRequestedArgs($request);
+                if ($op == 'download' && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 5) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                $canonicalUrlParams[] = $representationId;
+                $canonicalUrlParams[] = $assocId;
+                break;
+            case Application::ASSOC_TYPE_SUBMISSION:
+                $canonicalUrlOp = 'view';
+                if ($application->getName() == 'omp') {
+                    $canonicalUrlOp = 'book';
+                }
+                $canonicalUrlParams = [$submissionId];
+                $router = $request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($request);
+                $args = $router->getRequestedArgs($request);
+                if ($op == $canonicalUrlOp && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 3) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                break;
+            case Application::getContextAssocType():
+                $canonicalUrlOp = '';
+                $canonicalUrlPage = 'index';
+                break;
+        }
+        $canonicalUrl = $this->getCanonicalUrl($request, $canonicalUrlPage, $canonicalUrlOp, $canonicalUrlParams);
+
+        $fileType = null;
+        if (isset($mimetype)) {
+            $fileType = $this->getDocumentType($mimetype);
+        }
+
+        $userAgent = $request->getUserAgent();
+
+        // Retrieve the currently installed version
+        $versionDao = DAORegistry::getDAO('VersionDAO'); /** @var VersionDAO $versionDao */
+        $version = $versionDao->getCurrentVersion();
+        $versionString = $version->getVersionString();
```

Comment:
@asmecher, do you maybe have an idea how could I solve this here now?

Reply:
Hmm, until we have a more comprehensive entity caching approach I suggest we use the Registry to make this a one-time fetch. The only way this would ever change within an execution is during an upgrade, so make sure to check (`PKPApplication::isUpgrading`).

Reply:
I added Registry::set at the end of the PKPApplication::_construct(). I hope this is OK... ?

Reply:
I don't see it in the PR, but OK!

Reply:
Yes, I am pushing the changes now to another branch, to keep an overview :-)
S. https://github.com/bozana/pkp-lib/blob/6782-2/classes/core/PKPApplication.inc.php#L215-L219